scio-bigtable/src/main/java/com/spotify/scio/bigtable/BigtableDoFn.java

```diff
@@ -0,0 +1,301 @@
+  }
+
+  /**
+   * Create a {@link BigtableDoFn} instance.
+   * @param options Bigtable options.
+   * @param maxPendingRequests maximum number of pending requests to prevent runner from timing out
+   *                           and retrying bundles.
+   * @param cacheSupplier supplier for lookup cache.
+   */
+  public <K> BigtableDoFn(BigtableOptions options,
+                          int maxPendingRequests,
+                          CacheSupplier<A, B, K> cacheSupplier) {
+    this.instanceId = UUID.randomUUID();
+    this.options = options;
+    this.cacheSupplier = cacheSupplier;
+    this.semaphore = new Semaphore(maxPendingRequests);
+  }
+
+  @Setup
+  public void setup() {
+    LOG.info("Setup for {}", this);
+    session.computeIfAbsent(instanceId, instanceId -> {
+      LOG.info("Creating BigtableSession with BigtableOptions {}", options);
+      try {
+        return new BigtableSession(options);
+      } catch (IOException e) {
+        e.printStackTrace();
+        LOG.error("Failed to create BigtableSession", e);
+        throw new RuntimeException("Failed to create BigtableSession", e);
+      }
+    });
+    cache.computeIfAbsent(instanceId, instanceId -> cacheSupplier.createCache());
+  }
+
+  @StartBundle
+  public void startBundle() {
+    LOG.info("Start bundle for {}", this);
+    futures.clear();
+    results.clear();
+    errors.clear();
+    requestCount = 0;
+    resultCount = 0;
+  }
+
+  @SuppressWarnings("unchecked")
+  @ProcessElement
+  public void processElement(ProcessContext c, BoundedWindow window) {
+    flush(r -> c.output(KV.of(r.input, r.output)));
+
+    final A input = c.element();
+    B cached = cacheSupplier.get(instanceId, input);
+    if (cached != null) {
+      c.output(KV.of(input, cached));
+      return;
+    }
+
+    final UUID uuid = UUID.randomUUID();
+    ListenableFuture<B> future;
+    try {
+      semaphore.acquire();
+      try {
+        future = asyncLookup(session.get(instanceId), input);
+      } catch (Exception e) {
+        semaphore.release();
+        LOG.error("Failed to process element", e);
+        throw e;
+      }
+    } catch (InterruptedException e) {
+      LOG.error("Failed to acquire semaphore", e);
+      throw new RuntimeException("Failed to acquire semaphore", e);
+    }
+    requestCount++;
+
+    // Handle failure
+    Futures.addCallback(future, new FutureCallback<B>() {
+      @Override
+      public void onSuccess(@Nullable B result) {
+        semaphore.release();
+      }
+
+      @Override
+      public void onFailure(Throwable t) {
+        semaphore.release();
+        errors.add(t);
+        futures.remove(uuid);
+      }
+    });
+
+    // Handle success
+    ListenableFuture<B> f = Futures.transform(future, new Function<B, B>() {
+      @Nullable
+      @Override
+      public B apply(@Nullable B output) {
+        cacheSupplier.put(instanceId, input, output);
+        results.add(new Result(input, output, c.timestamp(), window));
```

Comment:
Could the result be emitted immediately using `c.output()` or is there some benefit to doing that later in `flush` ?

Reply:
The buffer behind `c.output()` is not thread safe, so everything has to be done on the main thread in `flush()`. Similar problem to what we had in Storm bolts.

Reply:
Offtopic: exactly five years later this question and comment helps other developers to achieve their goals. Thank you!